pkg/providers/tinkerbell/reconciler/reconciler.go

```diff
@@ -89,6 +91,19 @@ func (r *Reconciler) ValidateClusterSpec(ctx context.Context, log logr.Logger, c
 	return controller.Result{}, nil
 }
 
+// ValidateDatacenterConfig updates the cluster status if the TinkerbellDatacenter status indicates that the spec is invalid.
+func (r *Reconciler) RetrieveCAPIState(ctx context.Context, log logr.Logger, clusterSpec *c.Spec) (controller.Result, error) {
+	log = log.WithValues("phase", "retrieve CAPI state")
+
+	// Retrieve CAPI KubeadmControlPlane for control plane comparisons
+	// Retrieve CAPI MachineDeployments for worker node comparisons
+
+	// Store into ClusterApiState
+	r.clusterState = &tinkerbell.ClusterApiState{}
```

Comment:
what is the cluster api state here? I thought we're going to store operation in reconciler? like rolling/scaling/others

Reply:
State info is stored in Scope now